rosbag2_performance/rosbag2_benchmarking/rosbag2_benchmarking/raport_gen.py

```diff
@@ -0,0 +1,434 @@
+
+        # Calculate percent of msgs saved in bag
+        for topic, expected_msg_count in topic_msgs_accumulated.items():
+            val = self.bag_metadata[topic]['message_count']/expected_msg_count
+            self.result.bag_message_captured_percents.update(
+                {topic: val * 100}
+            )
+            self.result.bag_message_captured_num.update(
+                {topic: self.bag_metadata[topic]['message_count']}
+            )
+            self.result.bag_message_captured_expected.update(
+                {topic: expected_msg_count}
+            )
+
+        # CPU utilization
+        path = self.benchmark_path.joinpath('system_cpu.csv')
+        with open(path, newline='') as csvfile:
+            reader = csv.reader(csvfile, delimiter=';')
+            first_record = True
+            for row in reader:
+                self.result.cpu_utilization_t.append(int(row[0]))
+                cpus_number = len(row[2:])
+                if first_record:
+                    for i in range(2, cpus_number):
+                        self.result.cpu_utilization_v.append([])
+                    first_record = False
+
+                self.result.cpu_utilization_avg.append(float(row[1]))
+                for i in range(2, len(row)-1):
+                    self.result.cpu_utilization_v[i-2].append(float(row[i]))
+
+        # Disk utilization
+        path = self.benchmark_path.joinpath('system_disk.csv')
+        with open(path, newline='') as csvfile:
+            reader = csv.reader(csvfile, delimiter=';')
+            for row in reader:
+                self.result.disk_utilization.update(
+                    {int(row[0]): [float(x) for x in row[1:]]}
+                )
+                self.result.disk_utilization_t.append(int(row[0]))
+                self.result.disk_utilization_r.append(float(row[1])/1000)
+                self.result.disk_utilization_w.append(float(row[2])/1000)
+
+        # Memory utilization
+        path = self.benchmark_path.joinpath('system_mem.csv')
+        with open(path, newline='') as csvfile:
+            reader = csv.reader(csvfile, delimiter=';')
+            for row in reader:
+                self.result.mem_utilization.update(
+                    {int(row[0]): [float(x) for x in row[1:]]}
+                )
+                self.result.mem_utilization_t.append(float(row[0]))
+                self.result.mem_utilization_v.append(float(row[1])/1000000)
+
+        # Print report
+        self.result.base_time = self.bag_record_start_time
+        self.result.benchmark_path = self.benchmark_path
+        self.result.title = self.config['benchmark']['name']
+        self.result.generate_plots()
+        self.result.dump_report_yaml()
+        self.logger.info(str(self.result))
+
+
+def main():
+    """Ros2 once-spin run."""
+    rclpy.init()
+    node = ReportGen()
+    while rclpy.ok():
+        rclpy.spin_once(node, timeout_sec=0.1)
+        rclpy.shutdown()
+    node.destroy_node()
```

Comment:
why do we need a `rclpy` node for generating a report?

Reply:
You are correct, we should not need `rclpy` for that. It was used to keep the consistency in using running arguments (ie `--ros-args -p foo:=bar`). Will be fixed.